### PR TITLE
Allow log_momentum for adaptive optimizers

### LINKED ISSUE #5330: Logging momentum in LearningRateMonitor
## 🐛 Bug

Trying to log momentum in [LearningRateMonitor](https://pytorch-lightning.readthedocs.io/en/stable/generated/pytorch_lightning.callbacks.LearningRateMonitor.html?highlight=LearningRateMonitor) crashes training. When `log_momentum = False`, everything runs smoothly.

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/1JiSrB4JYexxMAjPPdnbv4Q0ShqRH2YJQ?usp=sharing

### Environment

* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.4
	- pyTorch_debug:     True
	- pyTorch_version:   1.7.0+cu101
	- pytorch-lightning: 1.1.2
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #5330

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes [if needed]?
- [x] Did you write any new necessary tests [no need for typos, docs]? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone are aligned!**
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/lr_monitor.py [hunk 1]
```diff
@@ -80,16 +80,28 @@ def on_train_start(self, trainer, *args, **kwargs):
...
 
+        if self.log_momentum:
+            def _check_momentum(key):
+                return any(
+                    key not in sch['scheduler'].optimizer.defaults for sch in trainer.lr_schedulers
+                )
+
+            if _check_momentum('momentum') and _check_momentum('betas'):
+                rank_zero_warn(
+                    "You have set log_momentum=True, but some optimizers doesn't"
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/lr_monitor.py:101]:
```suggestion
                    "You have set log_momentum=True, but some optimizers do not"
```

File: pytorch_lightning/callbacks/lr_monitor.py [hunk 2]
```diff
@@ -80,16 +80,28 @@ def on_train_start(self, trainer, *args, **kwargs):
...
+                    " have momentum. This will log 0 value for the momentum.", RuntimeWarning
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/lr_monitor.py:102]:
```suggestion
                    " have momentum. This will log a value 0 for the momentum.", RuntimeWarning
```

**Review by awaelchli — APPROVED** [review]
an idea for the future: allow the user to pass in functions / lambdas for extracting the learning rate, momentum or whatever required by the monitor. This way we can support all optimizers in a generic way and for the known ones, we provide the default functions. :)

File: tests/callbacks/test_lr_monitor.py [hunk 3]
```diff
@@ -69,7 +85,39 @@ def test_lr_monitor_single_lr_with_momentum(tmpdir):
...
+    lr_monitor = LearningRateMonitor(log_momentum=True)
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=1,
+        limit_val_batches=2,
+        limit_train_batches=5,
+        log_every_n_steps=1,
+        callbacks=[lr_monitor],
+    )
+    with pytest.warns(RuntimeWarning, match="optimizers doesn't have momentum."):
```
Comment (rohitgr7) [inline tests/callbacks/test_lr_monitor.py:112]:
```suggestion
    with pytest.warns(RuntimeWarning, match="optimizers do not have momentum."):
```

File: tests/callbacks/test_lr_monitor.py [hunk 4]
```diff
@@ -69,7 +85,39 @@ def test_lr_monitor_single_lr_with_momentum(tmpdir):
         'Expected momentum to be logged'
     assert len(lr_monitor.last_momentum_values) == len(trainer.lr_schedulers), \
         'Number of momentum values logged does not match number of lr schedulers'
-    assert all([k in ['lr-SGD-momentum'] for k in lr_monitor.last_momentum_values.keys()]), \
+    assert all([k in [f'lr-{opt}-momentum'] for k in lr_monitor.last_momentum_values.keys()]), \
```
Comment (carmocca) [inline tests/callbacks/test_lr_monitor.py:88]:
```suggestion
    assert all(k == f'lr-{opt}-momentum' for k in lr_monitor.last_momentum_values.keys()), \
```

File: tests/callbacks/test_lr_monitor.py [hunk 5]
```diff
@@ -69,7 +85,39 @@ def test_lr_monitor_single_lr_with_momentum(tmpdir):
...
+    )
+    with pytest.warns(RuntimeWarning, match="optimizers do not have momentum."):
+        result = trainer.fit(model)
+        assert result
+
+    assert all(v == 0 for v in lr_monitor.last_momentum_values.values()), \
+        'Expected momentum to be logged'
+    assert len(lr_monitor.last_momentum_values) == len(trainer.lr_schedulers), \
+        'Number of momentum values logged does not match number of lr schedulers'
+    assert all([k in ['lr-ASGD-momentum'] for k in lr_monitor.last_momentum_values.keys()]), \
```
Comment (carmocca) [inline tests/callbacks/test_lr_monitor.py:120]:
```suggestion
    assert all(k == 'lr-ASGD-momentum' for k in lr_monitor.last_momentum_values.keys()), \
```

File: pytorch_lightning/callbacks/lr_monitor.py [hunk 6]
```diff
@@ -80,16 +80,28 @@ def on_train_start(self, trainer, *args, **kwargs):
...
                 ' for `configure_optimizers` method.', RuntimeWarning
             )
 
+        if self.log_momentum:
+            def _check_momentum(key):
+                return any(
+                    key not in sch['scheduler'].optimizer.defaults for sch in trainer.lr_schedulers
+                )
+
+            if _check_momentum('momentum') and _check_momentum('betas'):
```
Comment (carmocca) [inline pytorch_lightning/callbacks/lr_monitor.py:99]: Better name

```suggestion
            def _check_no_key(key):
                return any(
                    key not in sch['scheduler'].optimizer.defaults for sch in trainer.lr_schedulers
                )

            if _check_no_key('momentum') and _check_no_key('betas'):
```

File: pytorch_lightning/callbacks/lr_monitor.py [hunk 7]
```diff
@@ -121,19 +133,25 @@ def _extract_stats(self, trainer, interval: str) -> Dict[str, float]:
...
-
                 else:
                     pg = param_groups[0]
                     lr = self._extract_lr(param_group=pg, name=name)
                     latest_stat.update(lr)
-                    momentum = self._extract_momentum(param_group=pg, name=f'{name}-momentum')
+                    momentum = self._extract_momentum(
+                        param_group=pg, name=f'{name}-momentum', use_betas=use_betas
+                    )
                     latest_stat.update(momentum)
```
Comment (carmocca) [inline pytorch_lightning/callbacks/lr_monitor.py:155]:
```suggestion
                for i, pg in enumerate(param_groups):
                    suffix = f'/pg{i + 1}' if len(param_groups) > 1 else ''
                    lr = self._extract_lr(param_group=pg, name=f'{name}{suffix}')
                    latest_stat.update(lr)
                    momentum = self._extract_momentum(
                        param_group=pg, name=f'{name}-momentum{suffix}', use_betas=use_betas
                    )
                    latest_stat.update(momentum)
```

**Review by tchaton — APPROVED** [review]
Outside of @carmocca comments, it looks good !

File: pytorch_lightning/callbacks/lr_monitor.py (hunk 7 above)
Comment (rohitgr7) [reply pytorch_lightning/callbacks/lr_monitor.py:155]: nice! will update in sometime.

File: pytorch_lightning/callbacks/lr_monitor.py [hunk 8]
```diff
@@ -33,11 +33,11 @@ class LearningRateMonitor(Callback):
     Automatically monitor and logs learning rate for learning rate schedulers during training.
 
     Args:
-        logging_interval: set to `epoch` or `step` to log `lr` of all optimizers
-            at the same interval, set to `None` to log at individual interval
+        logging_interval: set to ``'epoch'`` or ``'step'`` to log ``lr`` of all optimizers
+            at the same interval, set to ``None`` to log at individual interval
             according to the `interval` key of each scheduler. Defaults to ``None``.
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/lr_monitor.py:38]:
```suggestion
            according to the ``interval`` key of each scheduler. Defaults to ``None``.
```

File: tests/callbacks/test_lr_monitor.py [hunk 9]
```diff
@@ -69,7 +85,39 @@ def test_lr_monitor_single_lr_with_momentum(tmpdir):
         'Expected momentum to be logged'
     assert len(lr_monitor.last_momentum_values) == len(trainer.lr_schedulers), \
         'Number of momentum values logged does not match number of lr schedulers'
-    assert all([k in ['lr-SGD-momentum'] for k in lr_monitor.last_momentum_values.keys()]), \
+    assert all([k == f'lr-{opt}-momentum' for k in lr_monitor.last_momentum_values.keys()]), \
```
Comment (carmocca) [inline tests/callbacks/test_lr_monitor.py:88]: nit: you don't need to instantiate the list here, you can just use the generator. same for the tests below

```suggestion
    assert all(k == f'lr-{opt}-momentum' for k in lr_monitor.last_momentum_values.keys()), \
```

**Review by carmocca — APPROVED** [review]

File: tests/callbacks/test_lr_monitor.py [hunk 10]
```diff
@@ -69,7 +85,39 @@ def test_lr_monitor_single_lr_with_momentum(tmpdir):
...
+    )
+    with pytest.warns(RuntimeWarning, match="optimizers do not have momentum."):
+        result = trainer.fit(model)
+        assert result
+
+    assert all(v == 0 for v in lr_monitor.last_momentum_values.values()), \
+        'Expected momentum to be logged'
+    assert len(lr_monitor.last_momentum_values) == len(trainer.lr_schedulers), \
+        'Number of momentum values logged does not match number of lr schedulers'
+    assert all([k == 'lr-ASGD-momentum' for k in lr_monitor.last_momentum_values.keys()]), \
```
Comment (rohitgr7) [inline tests/callbacks/test_lr_monitor.py:120]:
```suggestion
    assert all(k == 'lr-ASGD-momentum' for k in lr_monitor.last_momentum_values.keys()), \
```

**Review by Borda — COMMENTED** [review]
lgtm 🐰
